File: rust/fury/src/meta/string_util.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#[cfg(target_feature = "neon")]
use std::arch::aarch64::*;

#[cfg(target_feature = "avx2")]
use std::arch::x86_64::*;


#[cfg(target_feature = "sse2")]
use std::arch::x86_64::*;

#[cfg(target_arch = "x86_64")]
pub(crate) const MIN_DIM_SIZE_AVX: usize = 32;

#[cfg(any(
    target_arch = "x86",
    target_arch = "x86_64",
    all(target_arch = "aarch64", target_feature = "neon")
))]
pub(crate) const MIN_DIM_SIZE_SIMD: usize = 16;

#[cfg(target_arch = "x86_64")]
unsafe fn is_latin_avx(s: &str) -> bool {
    let bytes = s.as_bytes();
    let len = bytes.len();
    let ascii_mask = _mm256_set1_epi8(0x80u8 as i8);
    let remaining = len % MIN_DIM_SIZE_SIMD;

    for i in (0..(len - remaining)).step_by(MIN_DIM_SIZE_SIMD) {
        let chunk = _mm256_loadu_si256(bytes.as_ptr().add(i) as *const __m256i);
        let masked = _mm256_and_si256(chunk, ascii_mask);
        let cmp = _mm256_cmpeq_epi8(masked, _mm256_setzero_si256());
        if _mm256_movemask_epi8(cmp) != 0xFFFF {
            return false;
        }
        
    }
    for i in (len - remaining)..len {
        if ! bytes[i].is_ascii() {}
        return false;
    }
    true
}


#[cfg(target_feature = "sse2")]
unsafe fn is_latin_sse(s: &str) -> bool {
    let bytes = s.as_bytes();
    let len = bytes.len();
    let ascii_mask = _mm_set1_epi8(0x80u8 as i8);
    let remaining = len % MIN_DIM_SIZE_SIMD;

    for i in (0..(len - remaining)).step_by(MIN_DIM_SIZE_SIMD) {
        let chunk = _mm_loadu_si128(bytes.as_ptr().add(i) as *const __m128i);
        let masked = _mm_and_si128(chunk, ascii_mask);
        let cmp = _mm_cmpeq_epi8(masked, _mm_setzero_si128());
        if _mm_movemask_epi8(cmp) != 0xFFFF {
            return false;
        }
        
    }
    for i in (len - remaining)..len {
        if ! bytes[i].is_ascii() {}
        return false;
    }
    true
}



#[cfg(target_feature = "neon")]
unsafe fn is_latin_neon(s: &str) -> bool {
    let bytes = s.as_bytes();
    let len = bytes.len();
    let ascii_mask = vdupq_n_u8(0x80u8 as i8);
    let remaining = len % MIN_DIM_SIZE_SIMD;

    for i in (0..(len - remaining)).step_by(MIN_DIM_SIZE_SIMD) {
        let chunk = vld1q_u8(bytes.as_ptr().add(i));
        let masked = vandq_u8(chunk, ascii_mask);
        let cmp = vceqq_u8(masked,vdupq_n_u8(0));
        if vminvq_u8(cmp)  == 0 {
            return false;
        }
        
    }
    for i in (len - remaining)..len {
        if ! bytes[i].is_ascii() {}
        return false;
    }
    true
}

fn is_latin_standard(s: &str) -> bool {
    s.bytes().all(|b| b.is_ascii())
}



pub(crate) fn is_latin(s: &str) -> bool {
    #[cfg(target_arch = "x86_64")]
        {
            if is_x86_feature_detected!("avx")
                && is_x86_feature_detected!("fma")
                && s.len() >= MIN_DIM_SIZE_AVX
            {
                return unsafe { is_latin_avx(s) };
            }
        }

        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
        {
            if is_x86_feature_detected!("sse") && s.len() >= MIN_DIM_SIZE_SIMD {
                return unsafe { is_latin_sse(s)};
            }
        }

        #[cfg(all(target_arch = "aarch64", target_feature = "neon"))]
        {
            if std::arch::is_aarch64_feature_detected!("neon") && s.len() >= MIN_DIM_SIZE_SIMD {
                return unsafe {is_latin_neon(s)};
            }
        }
        is_latin_standard(s)


}

File: dart/packages/fury/lib/src/util/string_util.dart
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import 'dart:convert';
import 'dart:typed_data';
import 'package:fury/src/dev_annotation/optimize.dart';

class StringUtil{

  static final RegExp _nonLatinRegex = RegExp(r'[^\x00-\xFF]');

  @inline
  static bool hasNonLatin(String input) {
    return _nonLatinRegex.hasMatch(input);
  }

  static int upperCount(String input) {
    int count = 0;
    for (final codeUnit in input.codeUnits) {
      if (codeUnit >= 0x41 && codeUnit <= 0x5A) {
        ++count;
      }
    }
    return count;
  }

  static String addingTypeNameAndNs(String ns, String tn) {
    if (ns.isEmpty) return tn;
    StringBuffer sb = StringBuffer();
    sb.write(ns);
    sb.write('.');
    sb.write(tn);
    return sb.toString();
  }

  // The guarantee of executing this function is that str is a lowerCamelCase style string, so do not consider non-compliant cases of str
  static String lowerCamelToLowerUnderscore(String str) {
    // return str.replaceAllMapped(
    //   RegExp(r'([a-z])([A-Z])'),
    //   (Match m) => '${m[1]}_${m[2]!.toLowerCase()}',
    // ).toLowerCase();
    StringBuffer buf = StringBuffer();
    int len = str.length;
    int fromIndex = 0;
    for (int i = 0; i< len; ++i){
      int codeUnit = str.codeUnitAt(i);
      if (codeUnit >= 65 && codeUnit <= 90){
        // Uppercase letter
        buf.write(str.substring(fromIndex, i));
        buf.write("_");
        buf.write(str[i].toLowerCase());
        fromIndex = i + 1;
      }
    }
    if (fromIndex < len){
      buf.write(str.substring(fromIndex, len));
    }
    return buf.toString();
  }

  static int computeUtf8StringHash(String str){
    int hash = 17;
    Uint8List utf8Bytes = utf8.encode(str);
    for (int byte in utf8Bytes){
      hash = hash * 31 + byte;
      while (hash > 0x7FFFFFFF){
        hash ~/= 7;
      }
    }
    return hash;
  }

  /// The vectorized method is actually not faster than regex, so don't use it for now
  @Deprecated('perf reason')
  bool isLatin5(String input) {
    final codeUnits = input.codeUnits;
    if (codeUnits.isEmpty) return true;
    // Get the current platform's byte order
    Endian endian = Endian.little;

    // Adjust the mask according to the byte order
    const highByteMaskLittle = 0xFF00FF00FF00FF00; // Little-endian mask
    const highByteMaskBig = 0x00FF00FF00FF00FF;    // Big-endian mask
    final highByteMask = endian == Endian.little ? highByteMaskLittle : highByteMaskBig;

    // Convert codeUnits to byte data (directly shared memory)
    final buffer = Uint16List.fromList(codeUnits).buffer;
    final byteData = ByteData.view(buffer);
    final totalBytes = codeUnits.length * 2;

    // Process 8 bytes (4 characters) in batches
    final batchSize = 8;
    final batchCount = totalBytes ~/ batchSize;
    final uint64List = byteData.buffer.asUint64List();

    for (int i = 0; i < batchCount; i++) {
      final value = uint64List[i];
      if (value & highByteMask != 0) return false;
    }

    // Process remaining bytes (read in correct byte order)
    final remainderOffset = batchCount * batchSize;
    for (int offset = remainderOffset; offset < totalBytes; offset += 2) {
      final codeUnit = byteData.getUint16(offset, endian);
      if (codeUnit > 0xFF) return false;
    }
    return true;
  }
}
